code/modules/projectiles/guns/energy/special.dm

```diff
@@ -757,5 +757,110 @@
+			var/obj/item/pinpointer/crew/C = locateUID(linked_pinpointer_UID)
+			C.linked_gun_UID = null
+			if(C.mode == MODE_DET)
+				C.stop_tracking()
+			linked_pinpointer_UID = null
+
+/obj/item/gun/energy/detective/proc/link_pinpointer(pinpointer_UID)
+	linked_pinpointer_UID = pinpointer_UID
+
+/obj/item/gun/energy/detective/multitool_act(mob/living/user, obj/item/I)
+	. = TRUE
+	user.visible_message("[user] starts [overcharged ? "restoring" : "removing"] the safety limits on [src].", "You start [overcharged ? "restoring" : "removing"] the safety limits on [src]")
+	if(!I.use_tool(src, user, 10 SECONDS, volume = I.tool_volume))
+		user.visible_message("[user] stops modifying the safety limits on [src].", "You stop modifying the [src]'s safety limits")
+		return
+	if(!overcharged)
+		overcharged = TRUE
+		ammo_type = list(/obj/item/ammo_casing/energy/detective/overcharge)
+		update_ammo_types()
+		select_fire(user)
+	else // Unable to early return due to the visible message at the end
+		overcharged = FALSE
+		ammo_type = list(/obj/item/ammo_casing/energy/detective, /obj/item/ammo_casing/energy/detective/tracker_warrant)
+		update_ammo_types()
+		select_fire(user)
+	user.visible_message("[user] [overcharged ? "removes" : "restores"] the safety limits on [src].", "You [overcharged ? "remove" : "restore" ] the safety limits on [src]")
+
+/obj/item/gun/energy/detective/attackby(obj/item/I, mob/user, params)
+	. = ..()
+	if(!istype(I, /obj/item/ammo_box/magazine/detective/speedcharger))
+		return
+	var/obj/item/ammo_box/magazine/detective/speedcharger/S = I
+	if(!S.charge)
+		to_chat(user, "<span class='notice'>[S] has no charge to give!</span>")
+		return
+	if(cell.charge == cell.maxcharge)
+		to_chat(user, "<span class='notice'>[src] is already at full power!</span>")
+		return
+	var/new_speedcharger_charge = cell.give(S.charge)
+	S.charge -= new_speedcharger_charge
+	S.update_icon()
+	update_icon()
+
+/obj/item/gun/energy/detective/process_fire(atom/target, mob/living/user, message, params, zone_override, bonus_spread)
+	if(overcharged)
+		if(prob(clamp((100 - ((cell.charge / cell.maxcharge) * 100)), 30, 80)))	//minimum probability of 30, maximum of 80
+			playsound(user, fire_sound, 50, 1)
+			visible_message("<span class='userdanger'>[src]'s energy cell overloads!</span>")
+			user.apply_damage(60, BURN, pick(BODY_ZONE_PRECISE_L_HAND, BODY_ZONE_PRECISE_R_HAND))
+			user.EyeBlurry(10 SECONDS_TO_LIFE_CYCLES)
+			user.flash_eyes(2, TRUE)
+			do_sparks(rand(5, 9), FALSE, src)
+			playsound(src, 'sound/effects/bang.ogg', 100, TRUE)
+			user.unEquip(src)
+			cell.charge = 0 //ha ha you lose
+			return
+	..()
+
+/obj/item/gun/energy/detective/proc/start_pointing(target_UID)
+	tracking_target_UID = target_UID
+	Announcer.autosay("Alert: Detective's revolver discharged in tracking mode. Tracking: [locateUID(tracking_target_UID)] at [get_area_name(src)].", src, "Security")
+	SEND_SIGNAL(src, COMSIG_DETGUN_TRACKING)
```

Comment:
is this signal really needed? surely you could just make this a proccall by getting a reference using `locateUID`. would make it is simpler for someone to understand your code.